openedx/core/djangoapps/user_api/management/commands/bulk_user_org_email_optout.py

```diff
@@ -0,0 +1,122 @@
+        )
+        parser.add_argument(
+            '--sleep_between',
+            default=self.DEFAULT_SLEEP_BETWEEN_DELETES,
+            type=float,
+            help='Seconds to sleep between chunked opt-outs.'
+        )
+        parser.add_argument(
+            '--optout_csv_path',
+            help='Filepath to CSV file containing user/org email opt-outs.'
+        )
+
+
+    def handle(self, *args, **options):
+        """
+        Execute the command.
+        """
+        dry_run = options['dry_run']
+        chunk_size = options.get('chunk_size', self.DEFAULT_CHUNK_SIZE)
+        sleep_between = options.get('sleep_between', self.DEFAULT_SLEEP_BETWEEN_DELETES)
+        optout_path = options['optout_csv_path']
+
+        if chunk_size <= 0:
+            raise CommandError('Only positive chunk size is allowed ({}).'.format(chunk_size))
+        if sleep_between < 0:
+            raise CommandError('Only non-negative sleep between seconds is allowed ({}).'.format(sleep_between))
+
+        # Read the CSV file. Log the number of user/org rows read.
+        with open(optout_path, 'r') as csv_file:
+            optout_reader = csv.reader(csv_file)
+            optout_rows = list(optout_reader)
+        log.info("Read %s opt-out rows from CSV file '%s'.", len(optout_rows), optout_path)
+
+        cursor = connections['default'].cursor()
+
+        # Update/insert the rows one chunk at a time.
+        curr_row_idx = 0
+        start_idx = 0
+        while curr_row_idx < len(optout_rows):
+            start_idx = curr_row_idx
+            end_idx = min(start_idx + chunk_size - 1, len(optout_rows) - 1)
+
+            log.info("Attempting opt-out for rows (%s, %s) through (%s, %s)...",
+                optout_rows[start_idx][0], optout_rows[start_idx][1],
+                optout_rows[end_idx][0], optout_rows[end_idx][1],
+            )
+
+            # Build the SQL query.
+            query = 'INSERT INTO user_api_userorgtag (`user_id`, `org`, `key`, `value`, `created`, `modified`) VALUES '
+            query_values = []
+            for idx in range(start_idx, end_idx + 1):
+                query_values.append('({},"{}","email-optin","False",NOW(),NOW())'.format(optout_rows[idx][0],optout_rows[idx][1]))
+            query += ','.join(query_values)
+            query += ' ON DUPLICATE KEY UPDATE value="False", modified=NOW();'
+
+            # Execute the SQL query.
+            if dry_run:
+                log.info(query)
+            else:
+                try:
+                    cursor.execute('START TRANSACTION;')
+                    cursor.execute(query)
+                except:
+                    cursor.execute('ROLLBACK;')
+                    log.error("Rolled-back opt-out for rows (%s, %s) through (%s, %s)!",
+                        optout_rows[start_idx][0], optout_rows[start_idx][1],
+                        optout_rows[end_idx][0], optout_rows[end_idx][1],
+                    )
+                    raise
```

Comment:
@macdiesel If there's an error in this chunk, the whole chunk will be rolled back. The script could continue here if preferred. There'd be a chunk of users to resolve either way - and the script could be run again on the entire CSV after resolution.